src/kernel/stage3.c

```diff
@@ -193,38 +218,51 @@ thunk create_init(kernel_heaps kh, tuple root, filesystem fs)
     return closure(heap_general(kh), startup, kh, root, fs);
 }
 
-closure_function(1, 1, status, kernel_read_complete,
-                 filesystem, fs,
+closure_function(2, 1, status, kernel_read_complete,
+                 filesystem, fs, boolean, destroy_fs,
                  buffer, b)
 {
-    add_elf_syms(b);
+    add_elf_syms(b, 0);
     deallocate_buffer(b);
-    destroy_filesystem(bound(fs));
+    if (bound(destroy_fs))
+        destroy_filesystem(bound(fs));
     closure_finish();
     return STATUS_OK;
 }
 
-closure_function(1, 2, void, bootfs_complete,
-                 kernel_heaps, kh,
+closure_function(4, 2, void, bootfs_complete,
+                 kernel_heaps, kh, tuple, root, boolean, klibs_in_bootfs, boolean, ingest_kernel_syms,
                  filesystem, fs, status, s)
 {
-    tuple root = filesystem_getroot(fs);
-    tuple c = children(root);
+    tuple boot_root = filesystem_getroot(fs);
+    tuple c = children(boot_root);
     assert(c);
-    table_foreach(c, k, v) {
-        if (!buffer_strcmp(symbol_string(k), "kernel")) {
-            kernel_heaps kh = bound(kh);
-            filesystem_read_entire(fs, v, heap_backed(kh),
-                                   closure(heap_general(kh),
-                                   kernel_read_complete, fs),
-                                   ignore_status);
-            break;
+    if (bound(klibs_in_bootfs)) {
+        init_klib(bound(kh), fs, boot_root);
+        if (table_find(bound(root), sym(klib_test))) {
+            load_klib("/klib/test", closure(heap_general(bound(kh)), klib_test_loaded));
+        }
```

Comment:
How about loading the module from within `init_klib()` instead? This way, all modules to be loaded are loaded from a single place (`init_klib()`), regardless of whether modules are in the boot or root filesystem. Otherwise, in general there is no safe place from which a module can be loaded, because of the asynchronous nature of filesystem mounting.

Commit https://github.com/nanovms/nanos/commit/cacc023798ffb8a70e2df3ffa0a9fb5bbbfb1557 from my (work in progress) https://github.com/nanovms/nanos/tree/feature/telemetry branch has a possible implementation for that.